samza-core/src/main/java/org/apache/samza/execution/JobNode.java

```diff
@@ -19,361 +19,150 @@
-    List<Long> windowTimerIntervals = operatorSpecs.stream()
-        .filter(spec -> spec.getOpCode() == OperatorSpec.OpCode.WINDOW)
-        .map(spec -> ((WindowOperatorSpec) spec).getDefaultTriggerMs())
-        .collect(Collectors.toList());
-
-    // Filter out the join operators, and obtain a list of their ttl values
-    List<Long> joinTtlIntervals = operatorSpecs.stream()
-        .filter(spec -> spec instanceof JoinOperatorSpec)
-        .map(spec -> ((JoinOperatorSpec) spec).getTtlMs())
-        .collect(Collectors.toList());
-
-    // Combine both the above lists
-    List<Long> candidateTimerIntervals = new ArrayList<>(joinTtlIntervals);
-    candidateTimerIntervals.addAll(windowTimerIntervals);
-
-    if (candidateTimerIntervals.isEmpty()) {
-      return -1;
+  // get all next operators consuming from the input {@code streamId}
+  Set<String> getNextOperatorIds(String streamId) {
+    if (!appDesc.getInputOperators().containsKey(streamId) || !inEdges.containsKey(streamId)) {
+      return new HashSet<>();
     }
-
-    // Compute the gcd of the resultant list
-    return MathUtil.gcd(candidateTimerIntervals);
+    return appDesc.getInputOperators().get(streamId).getRegisteredOperatorSpecs().stream()
+        .map(op -> op.getOpId()).collect(Collectors.toSet());
   }
 
-  /**
-   * This function extract the subset of configs from the full config, and use it to override the generated configs
-   * from the job.
-   * @param fullConfig full config
-   * @param generatedConfig config generated for the job
-   * @param configPrefix prefix to extract the subset of the config overrides
-   * @return config that merges the generated configs and overrides
-   */
-  private static Config extractScopedConfig(Config fullConfig, Config generatedConfig, String configPrefix) {
-    Config scopedConfig = fullConfig.subset(configPrefix);
-
-    Config[] configPrecedence = new Config[] {fullConfig, generatedConfig, scopedConfig};
-    // Strip empty configs so they don't override the configs before them.
-    Map<String, String> mergedConfig = new HashMap<>();
-    for (Map<String, String> config : configPrecedence) {
-      for (Map.Entry<String, String> property : config.entrySet()) {
-        String value = property.getValue();
-        if (!(value == null || value.isEmpty())) {
-          mergedConfig.put(property.getKey(), property.getValue());
-        }
-      }
+  InputOperatorSpec getInputOperator(String inputStreamId) {
+    if (!inEdges.containsKey(inputStreamId)) {
+      return null;
     }
-    scopedConfig = new MapConfig(mergedConfig);
-    log.debug("Prefix '{}' has merged config {}", configPrefix, scopedConfig);
-
-    return scopedConfig;
+    return appDesc.getInputOperators().get(inputStreamId);
   }
 
-  static String createId(String jobName, String jobId) {
-    return String.format("%s-%s", jobName, jobId);
+  boolean isLegacyTaskApplication() {
```

Comment:
Minor: would prefer to not proxy a util method from this class. Can this be done by the caller directly?

Reply:
Sounds good.